### DISCUSSION TITLE
end_offset in daily partition

### DISCUSSION THREAD

**Discussion by EtienneT** [dagster-io/dagster#14693] [Q&A] [answered]
I want to materialize a yearly partitionned asset using a daily partitionned asset that has an end_offset.  For example here I want to materialize price_yearly.  In this example price_final_daily always have a 5 days delays before the price settles.  So I can't materialize this asset recently.  price_recent_daily is the asset that contains the most recent prices which are not considered final, so there's no end_offset for this one.  price_yearly will get materialized each day and is simply a cached version of all pricing, including the most recent final and recent prices...

```python
@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-01-01", end_offset=-5)
)
def price_final_daily(context) -> pd.DataFrame:
    ...

@asset(
    partitions_def=DailyPartitionsDefinition(start_date="2023-01-01")
)
def price_recent_daily(context) -> pd.DataFrame:
    ...

yearly = TimeWindowPartitionsDefinition(
    cron_schedule="0 0 1 1 *",
    fmt="%Y-%m-%d",
    start="2023-01-01",
    end_offset=1,
)

@asset(
    partitions_def=yearly
)
def price_yearly(context, price_final_daily, price_recent_daily) -> pd.DataFrame:
    ...
```

But when I try to materialize price_yearly I get the following error:
`dagster._core.errors.DagsterInvalidInvocationError: Provided time windows [TimeWindow(start=DateTime(2023, 1, 1, 0, 0, 0, tzinfo=Timezone('UTC')), end=DateTime(2024, 1, 1, 0, 0, 0, tzinfo=Timezone('UTC')))] contain invalid time windows for partitions definition Daily, starting 2023-01-01 UTC. End offsetted by -5 partitions.`

It seems to try to include 5 daily partitions before the start_date which I don't have.

I tried using PartitionMapping like so:

```python
@asset(
    partitions_def=yearly,
    ins={
        "price_final_daily": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(
                start_offset=5
            ),
        )
    }
)
```

But to be able to use start_offset or end_offset of TimeWindowPartitionMapping, your upstream and downstream partition definition have to be the same which produce the following error:

`dagster._core.errors.DagsterInvalidDefinitionError: Can't use the start_offset or end_offset parameters of TimeWindowPartitionMapping when the cron schedule of the upstream PartitionsDefinition is different than the cron schedule of the downstream one. Attempted to map from cron schedule '0 0 1 1 *' to cron schedule '0 0 * * *'.`

What would be the best way to achieve this?

Thanks!

**Comment by sryza** [comment]
@clairelin135 - this seems relevant to what you've been working on recently.  They've got a yearly PartitionsDefinition whose last partition is "downstream" of days in the upstream daily PartitionsDefinitions that don't yet exist.

Would the fix be to use the `allow_nonexistent_upstream_partitions ` parameter that's being introduced here https://github.com/dagster-io/dagster/pull/14706?

If so, I think it's worth making the error message suggest this as a solution for people who encounter this in the future.

**Comment by EtienneT** [comment]
I have found the metadata `allow_missing_partitions` which I could set on my IOManager, but I just tried it and it doesn't do anything.

Tried setting it to all my assets to test:
```
@asset(
    partitions_def=yearly,
    metadata={
        'allow_missing_partitions': True
    }
)
```

But it doesn't do anything.

Even tried overriding _load_multiple_inputs in my iomanager, and forcing it there.  [This is where is is used in the code](https://github.com/dagster-io/dagster/blob/f627b3507ed95f109fb164d590678c5a09552c8f/python_modules/dagster/dagster/_core/storage/upath_io_manager.py#L179).

Here is my override, but it is not working:
```
class PandasFeatherIOManager(UPathIOManager):
    extension: str = ".feather"

    def _load_multiple_inputs(self, context: InputContext) -> Dict[str, Any]:
        context.metadata["allow_missing_partitions"] = True
        return super()._load_multiple_inputs(context)
```

If I put a context.log.info in there, it seems it is not even called before I get the error:

`dagster._core.errors.DagsterInvalidInvocationError: Provided time windows [TimeWindow(start=DateTime(2023, 1, 1, 0, 0, 0, tzinfo=Timezone('UTC')), end=DateTime(2024, 1, 1, 0, 0, 0, tzinfo=Timezone('UTC')))] contain invalid time windows for partitions definition Daily, starting 2023-01-01 UTC. End offsetted by -5 partitions.`

**Comment by EtienneT** [comment]
The only alternative I can think of is doing the offset inside the asset.  For the partition 2023-01-06, the partition data would actually contain the data from 2023-01-01.  Not ideal since the partition key would not correspond to the real range it is for, but this would circumvent this limitation where I can't specify that I want to allow un-materialized assets in the range.

**Comment by clairelin135** [comment, marked as answer]
Hi @EtienneT. In next week's release (1.3.10) we will ship an improvement to address this. On the `TimeWindowPartitionMapping` you'll be able to set a `allow_nonexistent_upstream_partitions` param that enables you to have nonexistent upstream partitions for a given partition. This way, you'll be able to kick off backfills and runs without the time window partition mapping throwing an error.
```
@asset(
    partitions_def=yearly,
    ins={
        "price_final_daily": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(allow_nonexistent_upstream_partitions=True)
        ),
        "price_recent_daily": AssetIn(
            partition_mapping=TimeWindowPartitionMapping(allow_nonexistent_upstream_partitions=True)
        ),
    },
)
def price_yearly(context, price_final_daily, price_recent_daily):
    ...
```

**Reply by EtienneT** [reply]
Thank you!
